File: system-test/typescript/src/testkit/dsl/core/scenario/given/given-promotion.ts
import { PromotionConfig } from '../scenario-context.js';
import { ThenContractStage } from '../then/then-contract.js';
import { WhenStage } from '../when/when-stage.js';
import type { GivenStage } from './given-stage.js';
import type { GivenPromotion as IGivenPromotion } from '../../../port/given/steps/given-promotion.js';

export class GivenPromotion implements IGivenPromotion {
  constructor(
    private readonly stage: GivenStage,
    private readonly config: PromotionConfig,
  ) {}

  withActive(promotionActive: boolean): this {
    this.config.promotionActive = promotionActive;
    return this;
  }

  withDiscount(discount: number | string): this {
    this.config.discount = typeof discount === 'number' ? discount.toFixed(2) : discount;
    return this;
  }

  and(): GivenStage {
    return this.stage;
  }

  when(): WhenStage {
    return this.stage.when();
  }

  then(): ThenContractStage {
    return this.stage.then();
  }
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/port/given/steps/GivenPromotion.java
package com.mycompany.myshop.backend.support.port.given.steps;

import com.mycompany.myshop.backend.support.port.given.steps.base.GivenStep;

public interface GivenPromotion extends GivenStep {
    GivenPromotion withActive(boolean promotionActive);

    GivenPromotion withDiscount(String discount);

    GivenPromotion withDiscount(double discount);
}
